### ISSUE TITLE
asyncssh chokes on ed25519 keys

### ISSUE THREAD

**Issue by blark** [ronf/asyncssh#100]
I have an ed25519 key in my .ssh direcotry and I get the following error. 

```
asyncssh.public_key.KeyImportError: Unknown OpenSSH private key algorithm
Unknown OpenSSH private key algorithm
```

Is there any way to disable reading of keys completely when initiating the connection? I just need to use username/pass authentication for this application. I guess, I also have the opposite question as well though, is there also a way to force use of a certain key and ONLY that key? 

Thanks!

**Comment by ronf** [comment]
AsyncSSH supports Ed25519 keys if you install Install libsodium from https://github.com/jedisct1/libsodium and libnacl from https://pypi.python.org/pypi/libnacl, so that would be one way to fix this. It would also give you support for curve25519 Diffie Hellman key exchange and the chacha20-poly1305 cipher.

Unfortunately, it looks like AsyncSSH will fail if you have a an ed25519 key installed under the default name ('.ssh/id_ed25519') but you don't have these packages installed. I should probably make that conditional on whether Ed25519 support is actually available or not.

As a temporary workaround, you could remove 'id_ed25519' from _DEFAULT_KEY_FILES in asyncssh/misc.py for now, if you don't want to install this support. It currently looks like:

```python
# Default file names in .ssh directory to read private keys from
_DEFAULT_KEY_FILES = ('id_ed25519', 'id_ecdsa', 'id_rsa', 'id_dsa')
```

Just change it to:

```python
# Default file names in .ssh directory to read private keys from
_DEFAULT_KEY_FILES = ('id_ecdsa', 'id_rsa', 'id_dsa')
```

**Comment by blark** [comment]
Ahh thanks, I did install some extra packages I must have missed that one. 

Any chance there is a way to do what I asked regarding authentication? 

1. Skip keys altogether only attempting PW auth?
2. Only attempt to auth with a specific key provided?

I don't see examples of this in the docs and I went through the source a bit but it's not immediately clear (sorry I'm a bit of a noob), thanks for your help!!

Mark

**Comment by ronf** [comment]
Yes. To skip public key authentication entirely, you can explicitly set the client_keys argument in create_connection() or connect() to None when you open the connection. If you want to load a specific key or set of keys instead of the default set in the .ssh directory (or the SSH agent), you can provide a key file or list of key files to this same argument. See more info at:

http://asyncssh.readthedocs.io/en/latest/api.html#specifyingprivatekeys

If you have an SSH agent set up, you may need to disable that as well if you don't want to use the keys it has in it. You can do that by explicitly setting the agent_path argument to these functions to None.

**Comment by blark** [comment]
Great thanks for your help!

**Comment by wookayin** [comment]
Just for someone who might find this issue in the future: As of asyncssh 1.13.0, it is `asyncssh.public_key` not `asyncssh.misc`.

A very dirty workaround way to monkey-patch the setting (I wish we had another way to ignore unsupported pubkeys:

```
# workaround of asyncssh + openssh issue (would work on asyncssh>=1.13)
# (e.g. Unknown OpenSSH private key algorithm)
try:
    if ('ssh-ed25519' not in asyncssh.public_key._public_key_algs) and \
            ('id_ed25519' in asyncssh.public_key._DEFAULT_KEY_FILES):
        asyncssh.public_key._DEFAULT_KEY_FILES = tuple(
            v for v in asyncssh.public_key._DEFAULT_KEY_FILES if v != 'id_ed25519'
        )
except:
    pass
```

**Comment by ronf** [comment]
Is the issue here that you want AsyncSSH to only add id_ed25519 as a default key when Ed25519 support is available? That seems reasonable, and I wouldn't mind making that behavior automatic.

As for patching stuff like this, any variables with a leading underscore are considered private to the module, and also anything in the package not exported explicitly from the top-level "asyncssh" module is also considered private, and subject to change from release to release. So, you're welcome to do this sort of thing if you need to, but I can't promise it won't break in a future release. You may need to make such a patch conditional on AsyncSSH version.

**Comment by ronf** [comment]
This change is now available as commit 333e08d in the "develop" branch. It now ignores any unknown key type errors which occur when loading keys from the default locations, similar to how it previously avoided returning an error when encrypted keys were found and no passphrase was specified.

**Comment by wookayin** [comment]
This is so great, thanks for your feedback and the change!
